src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Krivodonova.cpp

```diff
@@ -0,0 +1,96 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Krivodonova.hpp"
+
+#include <cmath>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace ScalarAdvection::Solutions {
+
+template <typename DataType>
+tuples::TaggedTuple<ScalarAdvection::Tags::U> Krivodonova::variables(
+    const tnsr::I<DataType, 1>& x, double t,
+    tmpl::list<ScalarAdvection::Tags::U> /*meta*/) const noexcept {
+  // map each grid points x(t) back to its initial position within [-1, 1] at
+  // t=0
+  auto x0 = make_with_value<tnsr::I<DataType, 1>>(x, 0.0);
+  get<0>(x0) = get<0>(x) - t;
+  for (size_t i = 0; i < get_size(get<0>(x)); ++i) {
+    auto& xi = get<0>(x0)[i];
+    xi = xi - 2.0 * floor(0.5 * (xi + 1.0));
```

Comment:
Sorry, I don't understand this piece of the code. It makes sense to me that line 25 remaps `x-> x- v * t` (with `v=1`). This looks like it's trying to map xi into a range (-1, 1), but I don't see how it can do that without knowing the size of the domain that the user put in the input file (which isn't available in the analytic solutions). I am also confused on the `floor` call and why that's needed.

You could add an `ASSERT` that each `x` is within `1.0` and `-1.0`. Something like `ASSERT(min(get<0>(x0)) >= -1.0 and max(get<0>(x0)) <= 1.0, "...");`

Reply:
@nilsdeppe Here is the logic: since we are using the periodic boundary condition with the domain [-1.0, 1.0], we need to do some kind of 'modulo' operation for x-vt, say,
```
x - vt = 2.0 * N + y    where N = integer & y is in [-1.0, 1.0) 
```
and get the value of `y` which will be the initial position of the point (x,t). Here 2.0 is nothing but the length of the 1D domain we are using.

So, we have
```
0.5 * (x - vt + 1.0) = N + 0.5 * (y+1.0)
``` 
Taking `floor` of this value grabs out the value of `N`, and removing 2N term from `x-vt` gives out the value `y`.

I worked this with assuming that the domain is always fixed with [-1.0, 1.0] -- I can add some more lines to deal with a generic size of domain, but not sure it would worth it if we will be using this solution just as a test

Reply:
Maybe I need to add these as comments!

Reply:
okay, that explains the math. I'm confused why you need this.  I think adding a comment would be good, and also clarifying that you need to deal with the modulo stuff because the domain is periodic :)